RFC #106: Wagtail 6.5* and 6.6* roadmap updates

thibaudcolas opened the pull request:
[View the RFC in HTML](https://github.com/thibaudcolas/rfcs/blob/106-roadmap-updates/text/106-roadmap-updates.md). Once reviewed with enough feedback and approved by the core team, those changes will go live on the [Wagtail Roadmap](https://wagtail.org/roadmap/).

6.5\* and 6.6\*: Provisional version numbers. See our [Release schedule](https://github.com/wagtail/wagtail/wiki/Release-schedule) for more info.

thibaudcolas commented:
This is now ready for review :)

thibaudcolas commented on text/106-roadmap-updates.md:
```diff
@@ -0,0 +1,93 @@
+
+### Headless API improvements
+
+Implementation of top 3-5 headless improvements as identified in the [2024 headless survey](https://wagtail.org/blog/2024-headless-survey/), with a particular focus on the built-in REST API and documentation.
```
LB: Make guide content available via API

Stormheg replied:
@thibaudcolas do you have any specific improvements in mind?

The RFC mentioned lists a couple specific improvements, but I find it hard to gauge how much effort they would take to make happen. I'm assuming we're going for low-effort, high-reward improvements.

thibaudcolas replied:
Here are the ones I have earmarked so far, which I selected from RFC 100, attempting to prioritise things that were high up on the list of features for people, and where we have existing clear improvements defined (some already with pull requests):

- New version of [headless support documentation](https://docs.wagtail.org/en/stable/advanced_topics/headless.html)
- [Fix headless Wagtail page URL #9984](https://github.com/wagtail/wagtail/pull/9984)
- [Improvements to API router reverse lookups #11310](https://github.com/wagtail/wagtail/pull/11310)
- [Retain fields parameter when redirecting from find_view of API #12141](https://github.com/wagtail/wagtail/pull/12141)
- [Refactor redirect middleware to use its lookup logic for redirects API #11524](https://github.com/wagtail/wagtail/pull/11524)
- [Override serializing StreamBlocks to JSON api #3454](https://github.com/wagtail/wagtail/issues/3454)
- [Documentation for using the "Dynamic image serve" view for generating image renditions on a headless site #6113](https://github.com/wagtail/wagtail/issues/6113)
- [API should provide access to the HTML-converted versions of rich text fields #2695](https://github.com/wagtail/wagtail/issues/2695)
- [No Image URL in the API for ImageChooserBlock #2087](https://github.com/wagtail/wagtail/issues/2087)
- Some work towards of [Support OpenAPI Schema generation for Wagtail API #6209](https://github.com/wagtail/wagtail/issues/6209)

Stormheg commented on text/106-roadmap-updates.md:
```diff
@@ -0,0 +1,93 @@
+
+The following Wagtail 6.4 roadmap items will be moved to Future:
+
+- [Autosave support refactorings with Telepath](https://github.com/wagtail/roadmap/issues/47)
```
**Suggestion:** perhaps make explicit that `Validation on publish` was identified during the 6.4 release cycle as something that takes precedence, which is why this was de-prioritized. 

**Question:** out of curiosity, does this feature have sponsors? Do any roadmap items have sponsors?

gasman replied:
Yes, we have a sponsor for validation-on-publish, and in fact it's constraints on the funding (and when budgets can be spent) that have led to this being prioritised above the Telepath work, rather than any technical need for it to come first.

thibaudcolas replied:
In the last release, drag'n'drop support was sponsored. In this upcoming release, in addition to validation on publish, we also have site settings permissions and some of the headless work (which bits exactly TBC)

Stormheg commented on text/106-roadmap-updates.md:
```diff
@@ -0,0 +1,93 @@
+
+For more information, see [RFC 104: Validation on publish](https://github.com/wagtail/rfcs/pull/104).
+
+### Headless userbar
```
**Suggestion:** built-in support for headless previews is a way more requested feature according to the 2024 survey than the userbar, change the title to reflect that.

```suggestion
### Headless preview & userbar
```

lb- replied:
Also. It's probably a good idea to also raise an issue in the current headless preview repo that we intend to merge/replace it. 

https://github.com/torchbox/wagtail-headless-preview

I know it's a Torchbox repo, but still good to add there, plus look at any outstanding issues/PRs that may indicate issues or features.

I assume that project would eventually become read only and maybe move to Wagtail deprecated. https://github.com/wagtail-deprecated

zerolab replied:
I agree, thank you @lb-.
As the sole maintainer of that package, I can tell it needs a tiny bit of TLC, but overall it has been mostly unchaged for a few years

thibaudcolas replied:
I chose "headless userbar" because that’s something we have no support for at all, while "headless preview" already exists. But happy to reword if it’s clearer like that :)

Stormheg commented on text/106-roadmap-updates.md:
```diff
@@ -0,0 +1,93 @@
+
+### Site settings permissions
+
+A new permission model for sites, where groups can be assigned permissions at the level of individual sites. This better maps organisational structures, compared to the current model of permissions being assigned at the level of the settings model rather than instances.
+
+For more information, see [RFC 105: Site settings permissions](https://github.com/wagtail/rfcs/pull/105).
```
Just reviewed the linked RFC, good improvement 👍

Stormheg commented on text/106-roadmap-updates.md:
```diff
@@ -0,0 +1,93 @@
+
+### CSP compatibility audit
+
+An up-to-date review of [CSP compatibility issues](https://github.com/wagtail/wagtail/issues/1288) in Wagtail, with the view to help resolve them all in future releases.
```
**Question:** to be clear; this is just about reviewing the current state of affairs? Not working of fixing the remaining issues? I'm asking because I thought the state of affairs was already known as 'very close to full compatibility' but I could be mistaken.

thibaudcolas replied:
Yep! Yes at a high level we know we’re close, but our [last feature-by-feature review](https://github.com/wagtail/wagtail/issues/7053) was 4 years ago. We need a clearer record as we now have CSP [on our roadmap](https://github.com/wagtail/roadmap/issues/92) and [as a GSoC project idea](https://github.com/wagtail/gsoc/blob/main/project-ideas.md#content-security-policy-compatibility). I guess aside from auditing I would like to see all problems spotted during the audit clearly logged in Wagtail issues, or logged to a "CSP compatibility" board in GitHub Projects.

And if there is time, I’d also like this to include incremental steps towards our [CSP roadmap item](https://github.com/wagtail/roadmap/issues/92):

- Possibly [updating the bakerydemo CSP](https://github.com/wagtail/bakerydemo/blob/main/.env.example) to be on by default, and document stricter options that are compatible, if possible.
- Possibly updating the developer docs to document current level of CSP support
- Possibly updating guide.wagtail.org or wagtail.org with a CSP

danielabrahamsgosh replied:
I found it difficult adding CSP to a Wagtail site around 2 years ago, especially if a strict policy was required. Even with `/admin` paths excluded or given a very liberal policy, the admin helper icon (bottom right hand of the window when logged in and viewing the front end) always caused issues.

I would suggest fixing the helper icon first so developers can at least fully decouple front end from back end and apply their CSP policy as they see fit. I had to override and customise it. Happy to share my implementation but am sure there are some hacks and idiosyncrasies yet to be overcome.

zerolab replied:
@danielabrahamsgosh please do share your overrides/customisations - I am certain we can learn from them and some of that. Even better, if you have the time, please tidy it up and open a PR?

thibaudcolas replied:
Makes a lot of sense to me! We can give that higher priority, as well as any other aspects of Wagtail living outside the admin area.

Stormheg commented on text/106-roadmap-updates.md:
```diff
@@ -0,0 +1,93 @@
+
+### Headless demo site
+
+An official headless demo site – primarily so our contributors and maintainers can more easily work on headless improvements.
+
+For more information, see [RFC 100: Enhancing headless support](https://github.com/wagtail/rfcs/pull/100).
```
**Question:** will this be a state-of-the-art site following all the latest best practices? Or will it be more of a reference implementation showcasing the (headless) features Wagtail provides.

**Question:** what's the plan for maintaining such a site? E.g. making sure it continues working and stays up-to-date with whatever is happening in JS-land.

thibaudcolas replied:
The primary goal is to be able to demo / test the headless capabilities of Wagtail in a representative site – so "reference implementation". Demonstrating best practices would be good, where it makes sense.

---

Re maintenance, we envisioned implementing the site to keep the maintenance work low: no hosted environment, make changes in the [existing bakerydemo](https://github.com/wagtail/bakerydemo) as much as possible (for example server-side API configuration), limit adoption of patterns that are very framework-specific where possible.

I foresee there will be more maintenance trouble than we’d want anyway, but in my mind this kind of investment is essential if we want to keep supporting headless site architectures.

For what it’s worth, here is the provisional planning done with @laymonage with the vue to repurpose my [bakerydemo-nextjs](https://github.com/thibaudcolas/bakerydemo-nextjs) as this new demo site:

- Keep Next.js, as it’s the most popular option per our survey. Use server-rendered Next.js as it will be a better test of headless Wagtail capabilities.
- Keep TypeScript. Keep using the existing bakerydemo’s styles and template markup wherever possible
- Switch from GraphQL to the built-in Wagtail REST API.

Stormheg reviewed (commented):
Looks like 6.5 is shaping up to be one for the benefit of headless users 👍 

Couple of questions, no objections

thibaudcolas commented:
Thank you for the feedback @Stormheg! There are a few elements of my answers I’ll want to move within the RFC contents but I’ll leave them as comments for now for others ’ benefit. I brought this to last week’s core team meeting briefly, and yesterday’s with a few more details, so marking this as "Final Comment Period". There’s lots of time for comments, I’ll be looking at updating our roadmap towards the second half of the month assuming we get enough feedback by then.

thibaudcolas commented:
I’ve updated the RFC based on feedback (thank you @Stormheg for the excellent questions!). Unfortunately I’ve also moved "Headless preview & userbar" from the Wagtail 6.5\* release to "Future" as I didn’t realise we’re not clear yet when this can go ahead. Could still be 6.5\* but it’s better for us to commit to fewer things that are more firm.

lb- commented:
Is it worth splitting the headless preview and headless userbar into two items?

Maybe it's possible to merge the headless preview package into core as a somewhat easier first step. This may help set up some momentum for headless usage, even if it's one less dependency.

thibaudcolas commented:
Good shout, thank you @lb-!

Stormheg reviewed (approved):
LGTM

thibaudcolas commented:
Thank you @Stormheg! Other feedback to date has only been positive so I’ll move forward with the corresponding public roadmap changes. If anyone has further feedback about any of the points, we can still make changes later.

thibaudcolas commented:
I have added a "sponsored" label to our [roadmap items in GitHub](https://github.com/wagtail/roadmap/issues?q=sort%3Aupdated-desc%20is%3Aissue%20label%3Asponsored%20) following @Stormheg’s question. That label is in addition to the existing "needs sponsorship" and "needs contributions" labels, to increase transparency a bit. We normally showcase sponsored work in the release notes – that label will make this same information more visible and available earlier.

I’ve spent a bit of time retroactively labelling past roadmap items. And added the label to our Google Summer of Code, Google Season of Docs, and Outreachy projects. Not sure if that’s helpful or no but "we" do receive funding for those in some ways so I guess that counts?

- GSoC projects are financially supported by Google, with a $1.5-3k stipend for contributors, and $500 for the Wagtail org
- Season of Docs (one project) was supported by Google, going entirely to our contributor.
- Our Outreachy participations were supported by Torchbox and the Outreachy general fund, with $7k stipends for contributors.
